agent/qrexec-agent-data.c

```diff
@@ -546,6 +572,226 @@ int process_child_io(libvchan_t *data_vchan,
     return child_process_status;
 }
 
+#define QUBES_MAX_SERVICE_NAME_LEN 31ULL
+#define QUBES_MAX_SERVICE_ARG_LEN 63ULL
+#define QUBES_MAX_SERVICE_DESCRIPTOR_LEN \
+    (QUBES_MAX_SERVICE_NAME_LEN + QUBES_MAX_SERVICE_ARG_LEN + 1ULL)
+#define QUBES_SOCKADDR_UN_MAX_PATH_LEN 1024
+
+#define MAKE_STRUCT(x) \
+    x("/usr/local/etc/qubes-rpc/") \
+    x("/etc/qubes-rpc/")
+static const struct Q {
+    const char *const string;
+    size_t const length;
+} paths[] = {
+#define S(z) { .string = z, .length = sizeof(z) - 1 },
+    MAKE_STRUCT(S)
+#undef S
+};
+static_assert(sizeof("/etc/qubes-rpc/") == 16, "impossible");
+static_assert(sizeof("/usr/local/etc/qubes-rpc/") == 26, "impossible");
+// static_assert(QUBES_MAX_SERVICE_DESCRIPTOR_LEN == 95,
+//         "bad macro definition");
+#define S(z) \
+    static_assert(sizeof(z) + QUBES_MAX_SERVICE_DESCRIPTOR_LEN <= QUBES_SOCKADDR_UN_MAX_PATH_LEN, \
+            "Path too long: " #z);
```

Comment:
Since no filename can be longer than `NAME_MAX`, no service name can (usefully) be either.  That is true for both executable and socket services.

Reply:
Remember that full service descriptor is in form of `service_name+argument` and only the `service_name` strictly needs to fit a path name (in the current implementation). It does mean that for some very long argument it may not be possible to create a service file/socket for this specific argument - which is fine as you can still handle it with a generic `service_name` file/socket that gets an argument as an actual argument. But this also means `service_name+argument` combination can be legally longer than `NAME_MAX`.

Reply:
You are correct.

My current plan is:

* Reject service names longer than `NAME_MAX` early on.  There is no point in even trying to execute them.
* If `service_name+argument` is longer than `NAME_MAX`, skip the search for a service file/socket for the specific argument, as it cannot succeed.